src/v/archival/ntp_archiver_service.cc

```diff
@@ -1555,4 +1571,29 @@ size_t ntp_archiver::get_local_segment_size() const {
     return log_segment_size;
 }
 
+ss::future<bool>
+ntp_archiver::prepare_transfer_leadership(ss::lowres_clock::duration timeout) {
+    _paused = true;
+
+    try {
+        co_await ss::get_units(_uploads_active, 1, timeout);
+        co_return true;
```

Comment:
Ah I understand -- the idea is that we're still leader here, but we know we're going to transfer leadership soon so we pause uploads.

Reply:
So IIUC this comes with the tradeoff that while we're under load, a leadership transfer away from a heavily-uploading node may take up to 5s at default. Not necessarily a dealbreaker, but something to take note of.

Reply:
> So IIUC this comes with the tradeoff that while we're under load, a leadership transfer away from a heavily-uploading node may take up to 5s at default. Not necessarily a dealbreaker, but something to take note of.

Right, it definitely needs release-noting.

I'm adding a configuration property for the timeout, with the option for setting it to null to disable this behavior entirely in case of issues.